src/Microsoft.Identity.Web/AuthorizeForScopesAttribute.cs

```diff
@@ -14,152 +14,152 @@
-            {
-                return FindMsalUiRequiredExceptionIfAny(exception.InnerException);
-            }
-            else
-            {
-                return null;
-            }
-        }
-
-        private static bool IsAjaxRequest(HttpRequest request)
-        {
-            return string.Equals(request.Query[Constants.XRequestedWith], Constants.XmlHttpRequest, StringComparison.Ordinal) ||
-                string.Equals(request.Headers[Constants.XRequestedWith], Constants.XmlHttpRequest, StringComparison.Ordinal);
-        }
-    }
+	/// <summary>
+	/// Filter used on a controller action to trigger incremental consent.
+	/// </summary>
+	/// <example>
+	/// The following controller action will trigger.
+	/// <code>
+	/// [AuthorizeForScopes(Scopes = new[] {"Mail.Send"})]
+	/// public async Task&lt;IActionResult&gt; SendEmail()
+	/// {
+	/// }
+	/// </code>
+	/// </example>
+	public class AuthorizeForScopesAttribute : ExceptionFilterAttribute
+	{
+		/// <summary>
+		/// Scopes to request.
+		/// </summary>
+		public string[]? Scopes { get; set; }
+
+		/// <summary>
+		/// Key section on the configuration file that holds the scope value.
+		/// </summary>
+		public string? ScopeKeySection { get; set; }
+
+		/// <summary>
+		/// Azure AD B2C user flow.
+		/// </summary>
+		public string? UserFlow { get; set; }
+
+		/// <summary>
+		/// Allows specifying an AuthenticationScheme if OpenIdConnect is not the default challenge scheme.
+		/// </summary>
+		public string? AuthenticationScheme { get; set; }
+
+		/// <summary>
+		/// Handles the <see cref="MsalUiRequiredException"/>.
+		/// </summary>
+		/// <param name="context">Context provided by ASP.NET Core.</param>
+		public override void OnException(ExceptionContext context)
+		{
+			if (context != null)
+			{
+				MsalUiRequiredException? msalUiRequiredException = FindMsalUiRequiredExceptionIfAny(context.Exception);
+				if (msalUiRequiredException != null &&
+					IncrementalConsentAndConditionalAccessHelper.CanBeSolvedByReSignInOfUser(msalUiRequiredException))
+				{
+					// the users cannot provide both scopes and ScopeKeySection at the same time
+					if (!string.IsNullOrWhiteSpace(ScopeKeySection) && Scopes != null && Scopes.Length > 0)
+					{
+						throw new InvalidOperationException(
+							string.Format(
+								CultureInfo.InvariantCulture,
+								IDWebErrorMessage.ProvideEitherScopeKeySectionOrScopes,
+								nameof(ScopeKeySection),
+								nameof(Scopes)));
+					}
+
+					// Do not re-use the property Scopes. For more info: https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/issues/273
+					string[]? incrementalConsentScopes;
+
+					// If the user wishes us to pick the Scopes from a particular config setting.
+					if (!string.IsNullOrWhiteSpace(ScopeKeySection))
+					{
+						// Load the injected IConfiguration
+						IConfiguration configuration = context.HttpContext.RequestServices.GetRequiredService<IConfiguration>();
```

Comment:
GetRequiredService will throw if the service isn't registered. I recommend changing it to GetService to avoid an unhelpful error.